includes/Checker/Checks/Enqueued_Scripts_Scope_Check.php

```diff
@@ -0,0 +1,276 @@
+ *
+ * @since 1.0.2
+ */
+class Enqueued_Scripts_Scope_Check extends Abstract_Runtime_Check implements With_Shared_Preparations {
+
+	use Amend_Check_Result;
+	use Stable_Check;
+	use URL_Aware;
+
+	/**
+	 * List of viewable post types.
+	 *
+	 * @since 1.0.2
+	 * @var array
+	 */
+	private $viewable_post_types;
+
+	/**
+	 * List of plugin scripts.
+	 *
+	 * @since 1.0.2
+	 * @var array
+	 */
+	private $plugin_scripts = array();
+
+	/**
+	 * Plugin script counter.
+	 *
+	 * @since 1.0.2
+	 * @var array
+	 */
+	private $plugin_script_count = array();
+
+	/**
+	 * Gets the categories for the check.
+	 *
+	 * Every check must have at least one category.
+	 *
+	 * @since 1.0.2
+	 *
+	 * @return array The categories for the check.
+	 */
+	public function get_categories() {
+		return array( Check_Categories::CATEGORY_PERFORMANCE );
+	}
+
+	/**
+	 * Runs this preparation step for the environment and returns a cleanup function.
+	 *
+	 * @since 1.0.2
+	 *
+	 * @return callable Cleanup function to revert any changes made here.
+	 *
+	 * @throws Exception Thrown when preparation fails.
+	 */
+	public function prepare() {
+		$orig_scripts = isset( $GLOBALS['wp_scripts'] ) ? $GLOBALS['wp_scripts'] : null;
+
+		// Backup the original values for the global state.
+		$this->backup_globals();
+
+		return function () use ( $orig_scripts ) {
+			if ( is_null( $orig_scripts ) ) {
+				unset( $GLOBALS['wp_scripts'] );
+			} else {
+				$GLOBALS['wp_scripts'] = $orig_scripts;
+			}
+
+			$this->restore_globals();
+		};
+	}
+
+	/**
+	 * Returns an array of shared preparations for the check.
+	 *
+	 * @since 1.0.2
+	 *
+	 * @return array Returns a map of $class_name => $constructor_args pairs. If the class does not
+	 *               need any constructor arguments, it would just be an empty array.
+	 */
+	public function get_shared_preparations() {
+		$demo_posts = array_map(
+			static function ( $post_type ) {
+				return array(
+					'post_title'   => "Demo {$post_type} post",
+					'post_content' => 'Test content',
+					'post_type'    => $post_type,
+					'post_status'  => 'publish',
+				);
+			},
+			$this->get_viewable_post_types()
+		);
+
+		return array(
+			Demo_Posts_Creation_Preparation::class => array( $demo_posts ),
+		);
+	}
+
+	/**
+	 * Runs the check on the plugin and amends results.
+	 *
+	 * @since 1.0.2
+	 *
+	 * @param Check_Result $result The check results to amend and the plugin context.
+	 */
+	public function run( Check_Result $result ) {
+		$urls = $this->get_urls();
+		$this->run_for_urls(
+			$urls,
+			function () use ( $result ) {
+				$this->check_url( $result );
+			}
+		);
+
+		if ( ! empty( $this->plugin_scripts ) ) {
+			$url_count = count( $urls );
+			foreach ( $this->plugin_scripts as $plugin_script ) {
+				if ( isset( $plugin_script['count'] ) && ( $url_count === $plugin_script['count'] ) ) {
+					$this->add_result_warning_for_file(
+						$result,
+						__( 'This script is being loaded in all contexts.', 'plugin-check' ),
```

Comment:
Why is it a problem if a script is loaded in all contexts?

Reply:
It's not _always_ a problem, but often times this is done in error and not actually on purpose. This can also be often observed in the wp-admin, when a plugin loads JS on every page instead of only its own settings page.

This check is aimed to detect the former, but of course it can't be bulletproof. Hence this is a _warning_ and not an _error_.